Write a tutorial for dates, schedules, calendars and dcf (#468)
digital-asset/docs.daml.com: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst

Since there is one year between the start and the end date, we would expect this periodic schedule
to correspond to four periods of 3M each:

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Schedule.daml
  :language: daml
  :start-after: -- CREATE_EXPECTED_SCHEDULE_RESULT_BEGIN
  :end-before: -- CREATE_EXPECTED_SCHEDULE_RESULT_END

Note that we distinguish between adjusted and unadjusted dates. The unadjusted date is the result of
adding a specified time period to the start date of the previous schedule period. This can fall on a
non-business day. By applying a
:ref:`BusinessDayAdjustment <type-daml-finance-interface-types-date-calendar-businessdayadjustment-93933>`,
we also get an adjusted date, which must fall on a business day.

Now, let us roll out this schedule to get the specific start and end date for each period, and
compare it to the expected result:

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Schedule.daml
  :language: daml
  :start-after: -- CREATE_SCHEDULE_BEGIN
  :end-before: -- CREATE_SCHEDULE_END

An important aspect of schedules is the concept of *stub periods*. The above schedule only had

> regular 3M periods. However, if the sample schedule would start one month later:

Reviewer: Not sure about "reveals what happens"

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Schedule.daml
  :language: daml
  :start-after: -- CREATE_PERIODIC_SCHEDULE_WITH_STUB_BEGIN
  :end-before: -- CREATE_PERIODIC_SCHEDULE_WITH_STUB_END

it would only have 11 months in total, which does not match with neither 3 nor 4 regular 3M periods.
Instead, we need to define a stub period, so that the rest of the schedule is regular.
In the example above, we did this by defining a start date for the first regular period. This

> implies a short initial stub period, in our case 2 months. We would expect the following schedule:

Reviewer: Do you maybe mean "assumes" rather than "implies"? Not sure.

> implies a short initial stub period, in our case 2 months. We would expect the following schedule:

Author: I meant that _defining a start date for the first regular period_ has the consequence a short initial stub period will be used, so that would be "implies", right?

which are supported using the
:ref:`DayCountConventionEnum <type-daml-finance-interface-types-date-daycount-daycountconventionenum-67281>`.

For example, consider the schedule with a short initial stub in the previous section. Assume we want
to use the ``Act360`` day count convention:

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/DayCount.daml
  :language: daml
  :start-after: -- CALCULATE_ACT360_DCF_PERIOD1_BEGIN
  :end-before: -- CALCULATE_ACT360_DCF_PERIOD1_END

which corresponds to the 2M short initial stub period. This day count fraction can be used to
calculate a bond coupon or a swap payment for the given period, by multiplying it with the yearly
interest rate of the instrument, e.g. 4% per annum.

We can compare this initial day count fraction with those of the 2nd and 3rd periods of this
schedule, which are both regular (3M):

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/DayCount.daml
  :language: daml
  :start-after: -- CALCULATE_ACT360_DCF_PERIOD2AND3_BEGIN
  :end-before: -- CALCULATE_ACT360_DCF_PERIOD2AND3_END

These day count fractions are clearly greater than the one of the first period, since they

> correspond to 3M instead of 2M. Note that they differ slightly, because they do not contain the

Reviewer: ```suggestion
correspond to 3M instead of 2M. Note that they differ slightly: they do not contain the
```

> exact same number of days.
> 
> In addition to ``Act360``, the
> :ref:`DayCountConventionEnum <type-daml-finance-interface-types-date-daycount-daycountconventionenum-67281>`
> also supports several other day count conventions. Some of them can be computed using
> :ref:`calcDcf <function-daml-finance-util-date-daycount-calcdcf-20432>`, using only a start and an
> end date as an input, as described above.

Reviewer: ```suggestion
exact same number of days.

In addition to ``Act360``, the
:ref:`DayCountConventionEnum <type-daml-finance-interface-types-date-daycount-daycountconventionenum-67281>`
supports several other day count conventions. You can compute some of them with
:ref:`calcDcf <function-daml-finance-util-date-daycount-calcdcf-20432>`, using only a start and an
end date as an input as shown above.
```

> Others, e.g. ``ActActISDA``, are a bit more complicated because they require additional information.
> They can be calculated using the
> :ref:`calcPeriodDcf <function-daml-finance-util-date-daycount-calcperioddcf-63067>` function
> instead, which takes a schedule period (containing stub information) as input, as well as the
> schedule end date and the payment frequency:

Reviewer: ```suggestion
Others, such as ``ActActISDA``, are a bit more complicated because they require additional information.
You can calculate them using the
:ref:`calcPeriodDcf <function-daml-finance-util-date-daycount-calcperioddcf-63067>` function
instead. This function takes a schedule period (containing stub information) as input, as well as the
schedule end date and the payment frequency:
```

Reviewer: I'd suggest adding some sort of conclusion or next steps. The page ends sort of abruptly.
